File: samples/dotnet-csharp-markdown/ContentService.cs
using Markdig;
using Microsoft.Graph.Models.ExternalConnectors;
using YamlDotNet.Serialization;

public interface IMarkdown
{
  string? Markdown { get; set; }
}

class BlogPost : IMarkdown
{
  [YamlMember(Alias = "title")]
  public string? Title { get; set; }
  [YamlMember(Alias = "slug")]
  public string? Slug { get; set; }
  [YamlMember(Alias = "excerpt")]
  public string? Excerpt { get; set; }
  [YamlMember(Alias = "image")]
  public string? Image { get; set; }
  [YamlMember(Alias = "date")]
  public DateTime? Date { get; set; }
  [YamlMember(Alias = "tags")]
  public IList<string>? Tags { get; set; }
  public string? Markdown { get; set; }
  public string? Content { get; set; }
}

static class ContentService
{
  static IEnumerable<BlogPost> Extract()
  {
    var blogPosts = new List<BlogPost>();

    var contentFolder = "content";
    var files = Directory.GetFiles(contentFolder);

    foreach (var file in files)
    {
      try
      {
        var contents = File.ReadAllText(file);
        var blogPost = contents.GetContents<BlogPost>();
        blogPost.Content = Markdown.ToPlainText(blogPost.Markdown ?? "");
        blogPosts.Add(blogPost);
      }
      catch (Exception ex)
      {
        Console.WriteLine(ex.Message);
      }
    }

    return blogPosts;
  }

  static IEnumerable<ExternalItem> Transform(IEnumerable<BlogPost> content)
  {
    var baseUrl = new Uri("https://blog.mastykarz.nl");

    return content.Select(a =>
    {
      return new ExternalItem
      {
        Id = a.Slug,
        Properties = new()
        {
          AdditionalData = new Dictionary<string, object> {
            { "title", a.Title ?? "" },
            { "excerpt", a.Excerpt ?? "" },
            { "imageUrl", new Uri(baseUrl, a.Image).ToString() },
            { "url", new Uri(baseUrl, a.Slug).ToString() },
            { "date", a.Date ?? DateTime.MinValue },
            { "tags@odata.type", "Collection(String)" },
            { "tags", a.Tags ?? new List<string>() }
          }
        },
        Content = new()
        {
          Value = a.Content ?? "",
          Type = ExternalItemContentType.Text
        },
        Acl = new()
        {
          new()
          {
            Type = AclType.Everyone,
            Value = "everyone",
            AccessType = AccessType.Grant
          }
        },
        Activities = new()
        {
          new()
          {
            OdataType = "#microsoft.graph.externalConnectors.externalActivity",
            Type = ExternalActivityType.Created,
            StartDateTime = a.Date,
            PerformedBy = new Identity
            {
              Type = IdentityType.User,
              Id = "9da37739-ad63-42aa-b0c2-06f7b43e3e9e"
            }
          }
        }
      };
    });
  }

  static async Task Load(IEnumerable<ExternalItem> items)
  {
    foreach (var item in items)
    {
      Console.Write(string.Format("Loading item {0}...", item.Id));
      try
      {
        await GraphService.Client.External
          .Connections[ConnectionConfiguration.ExternalConnection.Id]
          .Items[item.Id]
          .PutAsync(item);
        Console.WriteLine("DONE");
      }
      catch (Exception ex)
      {
        Console.WriteLine("ERROR");
        Console.WriteLine(ex.Message);
      }
    }
  }

  public static async Task LoadContent()
  {
    var content = Extract();
    var transformed = Transform(content);
    await Load(transformed);
  }
}

File: samples/java-markdown/app/src/main/java/connector/ContentService.java
package connector;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.commonmark.Extension;
import org.commonmark.ext.front.matter.YamlFrontMatterExtension;
import org.commonmark.ext.front.matter.YamlFrontMatterVisitor;
import org.commonmark.node.Node;
import org.commonmark.parser.Parser;
import org.commonmark.renderer.html.HtmlRenderer;

import com.google.gson.Gson;
import com.google.gson.JsonParser;
import com.microsoft.graph.externalconnectors.models.AccessType;
import com.microsoft.graph.externalconnectors.models.Acl;
import com.microsoft.graph.externalconnectors.models.AclType;
import com.microsoft.graph.externalconnectors.models.ExternalActivity;
import com.microsoft.graph.externalconnectors.models.ExternalActivityType;
import com.microsoft.graph.externalconnectors.models.ExternalItem;
import com.microsoft.graph.externalconnectors.models.ExternalItemContent;
import com.microsoft.graph.externalconnectors.models.ExternalItemContentType;
import com.microsoft.graph.externalconnectors.models.Identity;
import com.microsoft.graph.externalconnectors.models.IdentityType;
import com.microsoft.graph.externalconnectors.models.Properties;
import com.microsoft.graph.externalconnectors.requests.ExternalActivityCollectionPage;

class BlogPost {
    public String url;
    public String content;
    public Map<String, List<String>> metadata;

    private BlogPost() {
    }

    public static Builder builder() {
        return new Builder();
    }

    public static class Builder {
        private BlogPost post;

        private Builder() {
            post = new BlogPost();
        }

        public Builder content(String content) {
            post.content = content;
            return this;
        }

        public Builder metadata(Map<String, List<String>> metadata) {
            post.metadata = metadata;
            return this;
        }

        public BlogPost build() {
            return post;
        }
    }
}

public class ContentService {
    private static List<BlogPost> extract() throws IOException {
        final Path basePath = Paths.get(System.getProperty("user.dir"));
        final Path contentDir = basePath.resolve("../content");
        if (!Files.exists(contentDir)) {
            throw new FileNotFoundException(String.format("Content directory %s does not exist%n", contentDir));
        }

        final List<Path> filePaths = Files.walk(contentDir)
                .filter(Files::isRegularFile)
                .collect(Collectors.toList());

        List<Extension> extensions = Arrays.asList(YamlFrontMatterExtension.create());

        final Parser parser = Parser.builder()
                .extensions(extensions)
                .build();

        final List<BlogPost> blogPosts = new ArrayList<>();

        for (Path filePath : filePaths) {
            final String content = Files.readString(filePath);
            final Node document = parser.parse(content);
            YamlFrontMatterVisitor yamlVisitor = new YamlFrontMatterVisitor();
            document.accept(yamlVisitor);
            final Map<String, List<String>> yaml = yamlVisitor.getData();

            HtmlRenderer renderer = HtmlRenderer.builder().build();
            final String html = renderer.render(document);
            
            final BlogPost post = BlogPost.builder()
                    .content(html)
                    .metadata(yaml)
                    .build();

            blogPosts.add(post);
        }

        return blogPosts;
    }

    private static List<ExternalItem> transform(List<BlogPost> items) throws MalformedURLException {
        final List<ExternalItem> transformed = new ArrayList<>();
        final URL baseUrl = new URL("https://blog.mastykarz.nl");
        final DateTimeFormatter isoFormatter = DateTimeFormatter.ISO_DATE_TIME;
        final DateTimeFormatter localFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
        
        for (BlogPost item : items) {
            final Acl acl = new Acl();
            acl.accessType = AccessType.GRANT;
            acl.type = AclType.EVERYONE;
            acl.value = "everyone";

            final ExternalItemContent content = new ExternalItemContent();
            content.value = item.content;
            content.type = ExternalItemContentType.HTML;

            final LocalDateTime dateTime = LocalDateTime.parse(item.metadata.get("date").get(0), localFormatter);
            final Gson gson = new Gson();

            final Properties properties = new Properties();
            properties.additionalDataManager()
                .put("title", JsonParser.parseString(String.format("\"%s\"", item.metadata.get("title").get(0))));
            properties.additionalDataManager()
                .put("excerpt", JsonParser.parseString(String.format("\"%s\"", item.metadata.get("excerpt").get(0))));
            properties.additionalDataManager()
                .put("imageUrl", JsonParser.parseString(String.format("\"%s\"", new URL(baseUrl, item.metadata.get("image").get(0)).toString())));
            properties.additionalDataManager()
                .put("url", JsonParser.parseString(String.format("\"%s\"", new URL(baseUrl, item.metadata.get("slug").get(0)).toString())));
            properties.additionalDataManager()
                .put("date", JsonParser.parseString(String.format("\"%s\"", dateTime.format(isoFormatter).toString())));
            properties.additionalDataManager()
                .put("tags@odata.type", JsonParser.parseString(String.format("\"%s\"", "Collection(String)")));
            properties.additionalDataManager()
                .put("tags", gson.toJsonTree(item.metadata.get("tags")).getAsJsonArray());

            // uncomment and replace with your own user ID
            // to record the activity as performed by you
            //
            // final Identity createdBy = new Identity();
            // createdBy.type = IdentityType.USER;
            // createdBy.id = "9da37739-ad63-42aa-b0c2-06f7b43e3e9e";

            // final ZoneOffset zoneOffset = ZoneOffset.ofHours(2);

            // final ExternalActivity created = new ExternalActivity();
            // created.oDataType = "#microsoft.graph.externalConnectors.externalActivity";
            // created.type = ExternalActivityType.CREATED;
            // created.performedBy = createdBy;
            // created.startDateTime = dateTime.atOffset(zoneOffset);
            // final List<ExternalActivity> activities = Arrays.asList(created);
            
            final ExternalItem transformedItem = new ExternalItem();
            transformedItem.id = item.metadata.get("slug").get(0);
            transformedItem.content = content;
            transformedItem.acl = Arrays.asList(acl);
            transformedItem.properties = properties;
            // transformedItem.activities = new ExternalActivityCollectionPage(activities, null);

            transformed.add(transformedItem);
        }

        return transformed;
    }

    private static void load(List<ExternalItem> items) {
        for (ExternalItem item : items) {
            System.out.printf("Loading item %s...", item.id);
            try {
                GraphService.getClient().external()
                        .connections(ConnectionConfiguration.getExternalConnection().id)
                        .items(item.id)
                        .buildRequest()
                        .putAsync(item)
                        .get();
                System.out.println("DONE");
            } catch (Exception e) {
                System.out.println("ERROR");
                System.out.println(e.getMessage());
            }
        }
    }

    public static void loadContent() {
        try {
            final List<BlogPost> blogPosts = extract();
            final List<ExternalItem> transformed = transform(blogPosts);
            load(transformed);
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }
}
